py/dml/c_backend.py

```diff
@@ -917,120 +925,125 @@ def generate_implements(code, device):
-                out('ASSERT(data);\n')
-                simple_event_fun_deconstruct_data(method,
-                                                  '*(attr_value_t*)data');
-                out('SIM_attr_free((attr_value_t*)data);\n');
-                out('MM_FREE(data);\n');
-                out('}\n\n', preindent = -1)
-                splitting_point()
-
-                start_function_definition(
-                    f'attr_value_t {get_value_fun}'
-                    + '(conf_object_t *_obj, void *data)')
-                out('{\n', postindent = 1)
-                out('ASSERT(data);\n')
-                simple_event_fun_deconstruct_data(method,
-                                                  '*(attr_value_t*)data');
-                out('return SIM_attr_copy(*(attr_value_t*)data);\n');
-                out('}\n\n', preindent = -1)
-                splitting_point()
-
-                start_function_definition(
-                    f'void *{set_value_fun}'
-                    + '(conf_object_t *_obj, attr_value_t val)')
-                out('{\n', postindent = 1)
-                simple_event_fun_deconstruct_data(method, 'val');
-                out('attr_value_t *data = MM_MALLOC(1, attr_value_t);\n');
-                out('*data = SIM_attr_copy(val);\n');
-                out('return data;\n');
-                out('}\n\n', preindent = -1)
-                splitting_point()
+            start_function_definition(
+                f'lang_void *{info["set_value"]}'
+                + '(conf_object_t *_obj, attr_value_t val)')
+            out('{\n', postindent = 1)
+            out('_simple_event_data_t *out;\n')
+            out(('set_error_t error = _deserialize_simple_event_data(%s, '
+                 + '%d, %s, %s, &_id_info_ht, val, &out);\n')
+                % ((('(const uint32 []) {%s}'
+                     % (', '.join(map(str, method.dimsizes)),))
+                    if method.dimensions > 0 else 'NULL'),
+                   method.dimensions,
+                   (f'sizeof({info["args_type"].label}_t)'
+                    if method.inp else 0),
+                   (serialize.lookup_deserialize(info['args_type'])
+                    if method.inp else 'NULL'),
+                   ))
+            out('if (error != Sim_Set_Ok) {\n', postindent=1)
+            out('out = NULL;\n')
+            out('}\n', preindent=-1)
+            out('return (lang_void *) out;\n')
+            out('}\n\n', preindent = -1)
+            splitting_point()
 
     reset_line_directive()
 
+def generate_simple_events_control_methods():
+    start_function_definition(
+        'void _cancel_simple_events(conf_object_t *_obj, _identity_t domain)')
+    out('{\n', postindent = 1)
+    for (method, _) in list(simple_events.items()):
+        out('SIM_event_cancel_time('
+            + f'SIM_object_clock(_obj), {crep.get_evclass(method)}, _obj, '
+            + '_simple_event_predicate, (lang_void *) &domain);\n')
+    out('}\n\n', preindent = -1)
```

Comment:
I'm concerned this is _slow_. `SIM_event_cancel_time` has a bunch of overhead and is linear in the size of the event queue (_not_ just the number of events posted of the event class), and it's called for every after event class. If a user calls `cancel_after()` on multiple objects at the same time (like they'd do in reset operations), then that operation is `O(n * k * o)`, where `n`= total number of posted events, `k` = total number of after event classes, `o` = number of objects. (This is assuming `_simple_event_predicate` is `O(1)`--- although it is linear in the number of domains associated with the tested after event, that number is expected to be very low, typically just 1.)

Thing is, I don't know how we could improve upon this without either breaking checkpoint compatibility (by changing how after event classes are generated to reduce the number of them) or changing the very nice interface we currently have (by dropping `cancel_after()` being per-object in favor of a global method that can accept a list of objects.)

Reply:
I suppose the only potentially large number here is `o`; someone might run cancel_after on *all* objects. I expect both n and k to be small, and resets happen seldom.

One possible optimization would be to keep a cache of objects that have posted simple events. The cache is invalidated when event data is destroyed or deserialized. When the cache is invalid, cancel_after can rebuild it by scanning the queue once with e.g. SIM_event_find_next_time.